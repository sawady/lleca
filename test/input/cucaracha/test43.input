
fun fib(n : Int) : Int {
    if n <= 1 {
        res := n
    } else {
        res := fib(n - 2) + fib(n - 1)
    } 
    return res
}

fun probar(i : Int, n : Int) {
    if i <= n {
        putNum(fib(i))
        putChar(10)
        probar(i + 1, n)
    }
}

fun main() {
   probar(0, 20)
}
