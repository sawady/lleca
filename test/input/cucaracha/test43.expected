Cons(
  Function(
    fib,
    Cons(
      Parameter(
        n,
        Int
      ),
      Nil
    ),
    Int,
    Cons(
      StmtIf(
        ExprLe(
          ExprVar(
            n
          ),
          ExprConstNum(
            1
          )
        ),
        Cons(
          StmtAssign(
            res,
            ExprVar(
              n
            )
          ),
          Nil
        ),
        Just(
          Cons(
            StmtAssign(
              res,
              ExprAdd(
                ExprCall(
                  fib,
                  Cons(
                    ExprSub(
                      ExprVar(
                        n
                      ),
                      ExprConstNum(
                        2
                      )
                    ),
                    Nil
                  )
                ),
                ExprCall(
                  fib,
                  Cons(
                    ExprSub(
                      ExprVar(
                        n
                      ),
                      ExprConstNum(
                        1
                      )
                    ),
                    Nil
                  )
                )
              )
            ),
            Nil
          )
        )
      ),
      Cons(
        StmtReturn(
          ExprVar(
            res
          )
        ),
        Nil
      )
    )
  ),
  Cons(
    Function(
      probar,
      Cons(
        Parameter(
          i,
          Int
        ),
        Cons(
          Parameter(
            n,
            Int
          ),
          Nil
        )
      ),
      Unit,
      Cons(
        StmtIf(
          ExprLe(
            ExprVar(
              i
            ),
            ExprVar(
              n
            )
          ),
          Cons(
            StmtCall(
              putNum,
              Cons(
                ExprCall(
                  fib,
                  Cons(
                    ExprVar(
                      i
                    ),
                    Nil
                  )
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    10
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  probar,
                  Cons(
                    ExprAdd(
                      ExprVar(
                        i
                      ),
                      ExprConstNum(
                        1
                      )
                    ),
                    Cons(
                      ExprVar(
                        n
                      ),
                      Nil
                    )
                  )
                ),
                Nil
              )
            )
          ),
          Nothing
        ),
        Nil
      )
    ),
    Cons(
      Function(
        main,
        Nil,
        Unit,
        Cons(
          StmtCall(
            probar,
            Cons(
              ExprConstNum(
                0
              ),
              Cons(
                ExprConstNum(
                  20
                ),
                Nil
              )
            )
          ),
          Nil
        )
      ),
      Nil
    )
  )
)
